cpp/src/streaming/cudf/shuffler.cpp

```diff
@@ -10,37 +10,147 @@
+coro::task<void> insert_and_notify(
+    coro::mutex& mtx,
+    coro::condition_variable& cv,
+    std::unordered_set<shuffler::PartID>& set,
+    shuffler::PartID pid
 ) {
-    if (primary.value() != secondary.value()) {
-        RAPIDSMPF_CUDA_TRY(cudaEventRecord(event, secondary));
-        RAPIDSMPF_CUDA_TRY(cudaStreamWaitEvent(primary, event));
+    // Note: this coroutine is not needed to be scheduled, because it is called from the
+    // progress thread.
+    {
+        auto lock = co_await mtx.scoped_lock();
+        set.insert(pid);
     }
+    co_await cv.notify_all();
 }
 
 }  // namespace
 
+ShufflerAsync::ShufflerAsync(
+    std::shared_ptr<Context> ctx,
+    rmm::cuda_stream_view stream,
+    OpID op_id,
+    shuffler::PartID total_num_partitions,
+    shuffler::Shuffler::PartitionOwner partition_owner
+)
+    : ctx_(std::move(ctx)),
+      shuffler_(
+          ctx_->comm(),
+          ctx_->progress_thread(),
+          op_id,
+          total_num_partitions,
+          stream,
+          ctx_->br(),
+          [this](shuffler::PartID pid) -> void {
+              ctx_->comm()->logger().trace("inserting finished partition ", pid);
+              coro::sync_wait(insert_and_notify(mtx_, cv_, ready_pids_, pid));
+          },
+          ctx_->statistics(),
+          std::move(partition_owner)
+      ) {}
+
+bool ShufflerAsync::finished() const {
+    return shuffler_.finished();
+}
+
+void ShufflerAsync::insert(std::unordered_map<shuffler::PartID, PackedData>&& chunks) {
+    shuffler_.insert(std::move(chunks));
+}
+
+void ShufflerAsync::insert_finished(std::vector<shuffler::PartID>&& pids) {
+    shuffler_.insert_finished(std::move(pids));
+}
+
+coro::task<std::vector<PackedData>> ShufflerAsync::extract_async(shuffler::PartID pid) {
+    // Wait until the partition is finished
+    auto lock = co_await mtx_.scoped_lock();
+
+    RAPIDSMPF_EXPECTS(
+        !extracted_pids_.contains(pid),
+        "partition already extracted: " + std::to_string(pid),
+        std::out_of_range
+    );
+
+    co_await cv_.wait(lock, [this, pid]() {
+        // Note: purposefully not checking for extracted_pids_.contains(pid) here,
+        // because it would require notifying the cv every time a partition is extracted.
+        // Consequence of this is that, if pid was extracted by some other task, this task
+        // would only be notified during the shuffler.finished() check.
+        return shuffler_.finished() || ready_pids_.contains(pid);
+    });
+
+    // partition not found (may have been already extracted or shuffler was finished
+    // before the pid was inserted into ready_pids_)
+    RAPIDSMPF_EXPECTS(
+        !extracted_pids_.contains(pid) && ready_pids_.erase(pid) > 0,
+        "partition already extracted or not found: " + std::to_string(pid),
+        std::out_of_range
+    );
+    extracted_pids_.emplace(pid);
+    lock.unlock();  // no longer need the lock
+
+    auto chunks = shuffler_.extract(pid);
```

Comment:
This might fail (because `extract_any_async` can sneak in and steal the PID from us).

What about if we just drop the lock _after_ `shuffler_.extract(pid)` (and similarly in `extract_any_async`)? Then it is safe for `extract` and `extract_any_async` to run simultaneously I think.